[August] 미션 5: 보드에 위치 부여 및 점수 계산

polynomeer opened the pull request:
## 미션 5: 보드에 위치 부여 및 점수 계산

안녕하세요? 항상 친절한 리뷰 감사합니다. 이번 PR은 미션5에 관한 내용입니다. 미션1~4에 비해서 다소 어려운 난이도라고 느껴졌습니다. 그래서 한참을 고민하다가 중간중간에 자바지기님의 코드를 참조해가면서 단계별로 진행해보았습니다. 

특히 가장 어려운 부분이 Rank의 리스트를 관리하고 인덱스로 접근하는 부분이었습니다. 그냥 이차원배열로 한번에 선언하였다면 row와 column을 조정해가면서 바로 접근하면 되겠지만, Rank를 별도로 선언하니깐 그게 쉽지 않았습니다. Rank에는 별도의 인덱스를 인자로 넘겨주는 방식으로 해결한다는 점을 참고했습니다.

## 구현 내용

- Board

  - `countPieces()` : Board 전체에서 Color와 Type을 인자로 받아서 해당 플레이어의 해당 타입인 Piece가 몇 개 남았는지 총 개수를 리턴합니다.
  - `initializeEmpty()` : Board 전체를 빈칸으로 채웁니다.
  - `initialize()` : Boad에 체스게임 시작 시점으로 모든 Piece를 배치합니다. 

  - `findPiece()` : 문자열 좌표값이 주어졌을 때 해당 Piece의 인스턴스를 얻어올 수 있습니다.
  - `move()`: Piece와 Position(에 대한 문자열)을 인자로 받아서 Position에 해당하는 Y좌표(row)에 해당하는 Rank를 찾고, 그 Rank에서 또 다시 Rank::move를 호출하여 X좌표(column)에 대한 이동을 set합니다.
  - `calculatePoint()`: findPiecesByColor()를 통해서 해당 색깔 플레이어의 Piece의 리스트를 얻어오고, 각 Piece에 해당하는(enum에 정의된) 점수를 모두 더해서 리턴합니다.
  - `findPiecesByColor()`: 모든 Rank를 돌면서 Rank::findPiecesByColor()을 호출하면서 해당 색깔의 Piece를 모두 리스트에 담아서 리턴합니다.
  - `toString()`: 이전과 동일합니다.

- Game
  - 우선은 `showScore()`과 `showScoreByPiece()`를 게임이 시작하면 한번 호출하도록 작성하였는데, 추후 미션에서 게임을 시작한 이후의 로직을 별도로 분리할 계획입니다.

- Rank
  - `move()`: 해당 Rank(row)내부에서 X좌표(column)에 대한 이동을 set하도록 값을 수정합니다.
  - `countPieces()`: 해당 Rank에서 Color와 Type을 인자로 받아서 해당 색깔 플레이어의 해당 타입의 Piece가 몇 개 남았는지를 리턴합니다.
  - `initializeWhitePieces()` : 팩토리 메서드로 해당 Rank의 각 위치에 흰색 플레이어의 Piece를 배치합니다.
  - `initializeBlackPieces()` : 팩토리 메서드로 해당 Rank의 각 위치에 검정색 플레이어의 Piece를 배치합니다.
  - `initializeWhitePawns()` : 팩토리 메서드로 Pawn을 해당 Rank에 8개 모두 배치합니다.
  - `initializeBlackPawns()` : 팩토리 메서드로 Pawn을 해당 Rank에 8개 모두 배치합니다.
  - `initializeBlankLine()` : 해당 Rank를 모두 공백으로 채웁니다.
  - `findPiece()` : X좌표를 인자로 받아서 리스트의 해당 인덱스에 위치한 Piece를 리턴합니다.
  - `findPiecesByColor()` : Color를 인자로 받아서 해당 Rank 내부에서 해당 색깔 플레이어의 모든 Piece를 담아서 리턴합니다. 이때 이 리스트가 외부에서 변경되는 것을 막아주기 위해서 `Collections.unmodifialbeList`로 제한하였습니다.
  - `toString()` : 기존과 동일합니다.

- Color - 색깔의 종류를 갖는 enum 입니다.

- Type - 타입의 종류를 갖는 enum입니다. 각 타입은 표기문자와 기본점수를 갖고 있으며, `getDefaultPoint()`를 통해서 해당 타입에 대한 점수를 얻어올 수 있습니다.

- Position
  - 생성자는 문자열로 입력 받았을 때와 정수값으로 입력 받았을 때 모두 생성 가능하도록 오버라이딩 하였습니다.
  - 멤버필드의 x,y에는 1~8이 저장되도록 하였으므로, 실제 배열의 인덱스를 가져올 수 있도록 하는 메서드인 `getXIndex()` 와 `getYIndex()`를 추가하였습니다.
  - Position객체가 Piece를 생성할 때 마다 생성자를 호출하면서 지정되므로, 같은 위치일 때 같은 객체라고 판단하도록 equals와 hashCode를 오버라이딩 하였습니다.
  - `getPrevNextRows()` : 해당 Position의 같은 X좌표에서 인접한 앞,뒤의 행(Y좌표, row, rank)의 Position을 리스트로 리턴합니다. (인접한 Pawn이 있을 경우, 0.5점을 감소시킨다는 조건 때문에 필요한 메서드입니다.)

- PieceCreator
  - 각 Piece를 생성하는 팩토리 메서드를 모두 별도로 분리한 클래스 입니다.

- Piece
  - `getRepresentation()` : 색깔에 따라 적절한 타입의 표기법을 리턴합니다.
  - `verifyPiece()` : 색깔과 타입을 검증합니다.
  - `matchColor()` : 색깔이 일치하는지 확인합니다.
  - `matchType()` : 타입이 일치하는지 확인합니다.
  - `addPiecesByColor()` : 색깔이 일치할때만 인자로 받은 리스트에 Piece객체 자신을 추가합니다.
  - `getPoint()` : Pawn이 아니면 해당 Piece에 대한 점수를 Type의 enum에 정의된 getDefaultPoint()로 부터 얻어와서 리턴합니다. `Pawn이 세로로 인접해 있을 때 각 Pawn의 점수가 0.5점이 된다.`라는 조건이 정말 어려웠는데, 이런식으로 구현할 수 있구나.. 하면서 자바지기님의 코드를 참고했습니다.
  - Piece 또한 VO로서 같은 타입과 같은 색깔일 때 서로 같은 Piece임을 보장하기 위해서 equals와 hashCode를 오버라이딩 해주었습니다.
  - `compareTo()` :각 Piece를 점수별로 정렬하도록 구현하라는 요구사항이 있어서 Comparable을 구현하도록 하여 `compareTo()`를 점수에 대한 비교구문으로 정의하였습니다.

- PrintUtils
  - `showScore()` : Board::calculatePoint()를 사용하여 각 색깔의 플레이어 별로 점수를 출력합니다.
  - `showScoreByPiece()` : `Board::findPiecesByColor()`를 사용하여 점수에 대한 내림차순으로 각 색깔의 플레이어가 가지고 있는 Piece의 종류와 점수를 출력합니다.

https://sourcegraph.com/github.com/Polynomeer/java-chess@feature/position

polynomeer committed:
feat: Add Color and Type enum in Piece

Deleted all final constants in Piece class.
Added Color and Type enum in Piece class to substitute the constants.
Changed all parameters and codes which are dependent on the constants to use enum.

Added test codes in PieceTest class to test the methods getWhiteRepresentation() and getBlackRepresentation() in Piece::Type enum.

polynomeer committed:
feat: Add createBlank method in Piece class

Added createBlank() method in Piece class.
Modified create_piece() and verifyPiece() methods in PieceTest to adapt enum.

polynomeer committed:
refactor: Add createWhite and createBlack methods in Piece

Added createWhite and createBlack methods in Piece class to avoid code duplication.
Changed member field to final to prevent from being changed.

polynomeer committed:
refactor: Add Rank class to represent 2d board

Added Rank class and made board have list of them
to represent 2d board.
Separated roles in Board class which are related to pieces into Rank class.

polynomeer committed:
feat: Add Rank index and findPiece logic

Added index of Rank to find Piece by index.
Added findPiece() method in Board class.
Added equals() and hashCode() in both Position and Piece class to identify the same piece in the same position.
Added test code for findPiece().

polynomeer committed:
feat: Add move and initializeEmpty in Board class

Added move() and initializeEmpty() in Board class
to set the position want to locate.
Added move() in Rank class to set the position.

polynomeer committed:
feat: Add calculate point logic

Added calculatePoint() and findPiecesByColor() in Board class to calculate each point of piece.
Added findPiecesByColor in Rank class to pick piece by color in a Rank.
Added double point in Type enum to allocate the point for each piece.
Added methods related to matching color and type to verify pieces.

polynomeer committed:
refactor: Delete unused code

Moved toString() block to the end of code.
Deleted unused constructor and field.

polynomeer committed:
refactor: Separate Color and Type enum

Separated Color and Type enum into independent files.

polynomeer committed:
refactor: Add PieceCreator class

Added PieceCreator class to separate factory methods.

polynomeer committed:
refactor: Rename addPiece to add

Rename addPiece() to add() to simplify

polynomeer committed:
Fix: Fix errors due to dependencies on PieceCreator class

Fixed errors in test codes depending on the factory methods in PieceCreator class.

polynomeer committed:
feat: Add compareTo method in Piece class

Made Piece implement Comparable.
Added compareTo() method in Piece class to sort piece by point.
Added showScore() and showScoreByPiece() to print sorted score of each piece and total score.

polynomeer committed:
refactor: Add PrintUtils to separate print methods

Added PrintUtils to separate print methods from Game class.

polynomeer committed:
refactor: Rename package name from pieces to piece

Renamed package name from pieces to piece.
Added private constructor in PrintUtils to prevent from being instance.

polynomeer committed:
feat: Add PrintUtilsTest and test codes

Added test codes in PrintUtilsTest to test sorting pieces by point.

polynomeer committed:
chore: Aligned code to clarify

polynomeer committed:
refactor: Rename some method and variables

Rename sizeofPieces() to countPiece() to clarify the meaning of counting.
Rename nextColumn() to getPrevNextRows() to emphasize the meaning of adjacent rows.
Rename getXPost() and getYPos() to getXIndex() and getYIndex() to emphasize the meaning of 'index'.
Deleted size() method. (not used)

polynomeer committed:
refactor: Set List from getPrevNextRows() to unmodifiable

Set the List<Position> from the return value by getPrevNextRows() to unmodifiableList to prevent from being changed by outside causes.

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,129 +1,78 @@
-            addBlackPawn(Piece.createBlackPawn());
+    public double calculatePoint(Color color) {
+        List<Piece> pieces = findPiecesByColor(color);
+        double point = 0.0;
+        for (Piece piece : pieces) {
+            point += piece.getPoint(pieces);
```
`.stream()` 기반으로 개선해주세요.

polynomeer replied:
pieces 리스트의 stream을 얻어서 합계를 구하는 방식으로 변경해보았습니다.

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,129 +1,78 @@
-            addWhitePawn(Piece.createWhitePawn());
-            addBlackPawn(Piece.createBlackPawn());
+    public double calculatePoint(Color color) {
+        List<Piece> pieces = findPiecesByColor(color);
```
아래 합계 계산하는 로직과 한 스트림으로 묶이면 좋겠네요.

```java
findPiecesByColor(color).stream().map....

polynomeer replied:
묶으려고 했다가 `getPoint()`를 Board에 포함하게 되면서 그냥 변수로 두게 되었습니다!

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,129 +1,78 @@
-                .append(appendNewLine(getWhitePawnsResult()))
-                .append(appendNewLine(getWhitePiecesResult()));
-
+        ListIterator<Rank> rankIter = board.listIterator(board.size());
+        while (rankIter.hasPrevious()) {
+            sb.append(rankIter.previous().toString());
+        }
```
`.toString()` 에 사이드 이펙트 발생 우려가 있는 복잡한 로직은 넣지 말아주세요. 
그리고 `.toString()` 이 생성하는 문자열은 특정 로직에서 사용될 문자열이 아닌, 객체에 대한 범용적인 상태 표현이어야 합니다. 
콘솔에 보드의 상태를 나타내는 목적으로 사용할 로직은 별도의 메소드를 설계해서 빼 주시고 `.toString()` 에는 가급적 IDE가 자동 생성해주는 정도의 로직을 넣어 주면 좋겠습니다.

polynomeer replied:
`.toString()`의 목적에만 신경을 쓰고 사이드 이펙트를 고려하지 못 했습니다. `PrintUtils::showBoard()`로 이동했습니다.

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,129 +1,78 @@
-            sb.append(piece.getRepresentation());
+    public List<Piece> findPiecesByColor(Color color) {
+        ArrayList<Piece> pieces = new ArrayList<>();
+        for (Rank rank : board) {
```
여기도 `this.board` 의 원소를 `List<Piece>` 로 바꾼 다음 `.collect()` 하는 로직으로 재설계하면 좋겠네요.

polynomeer replied:
stream을 사용한 코드로 변경했습니다.

wheejuni commented on src/main/java/chess/Rank.java:
```diff
@@ -0,0 +1,112 @@
+import static piece.PieceCreator.*;
+import static util.StringUtils.appendNewLine;
+
+class Rank {
```
접근 제어자는요?

polynomeer replied:
패키지 내부에서만 접근하므로 디폴트로 두었습니다. 하지만 `StringUtils::showBoard()`에서 사용하게 되어서 public으로 변경하였습니다.

polynomeer replied:
저는 우선은 private으로 최소한의 접근 제한자를 두려고 했습니다. (최소한의 접근 제어자로 시작하는것이 좋다는 호눅스의 팁을 빌려서) 하지만 테스트할 필요가 있으며 클래스 외부에서 사용하는 경우에는 public과 package-private(default) 두 가지 중에 선택을 해야했습니다. 저는 그래도 패키지 내부에서만 사용한다면 default로 선언하는게 맞다고 생각했는데, 그게 문제가 되나요? 혹시 default는 실무에서 잘 사용하지 않는건가요? 궁금합니다!

wheejuni commented on src/main/java/chess/Rank.java:
```diff
@@ -0,0 +1,112 @@
+
+    private final List<Piece> pieces = new ArrayList<>(BOARD_SIZE);
+
+    void add(Piece piece) {
```
여기도요?

polynomeer replied:
이것은 클래스 내부에서만 사용하는데 실수를 했군요! private으로 변경했습니다.

wheejuni commented on src/main/java/chess/Rank.java:
```diff
@@ -0,0 +1,112 @@
+        pieces.add(piece);
+    }
+
+    List<Piece> getPieces() {
```
여기도 없네요? ㅠㅠ

polynomeer replied:
몇 개는 정말로 접근 제어자를 빠뜨린 부분이 있어서 주의해야 겠다는 생각이 들었습니다 ㅠㅠ

wheejuni commented on src/main/java/chess/Rank.java:
```diff
@@ -0,0 +1,112 @@
+        pieces.set(xPos, piece);
+    }
+
+    int countPieces(Color color, Type type) {
```
헐~ 여기도 ㅠㅠㅠ

polynomeer replied:
Board에서 사용하므로 public으로 변경했습니다! 이런 경우에는 좀 헷갈리네요.

wheejuni commented on src/main/java/chess/Rank.java:
```diff
@@ -0,0 +1,112 @@
+        return countPiece;
+    }
+
+    static Rank initializeWhitePieces(int rankIndex) {
```
여기도요.

polynomeer replied:
public으로 변경했습니다.

wheejuni commented on src/main/java/chess/Rank.java:
```diff
@@ -0,0 +1,112 @@
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        for (Piece piece : getPieces()) {
+            sb.append(piece.getRepresentation());
+        }
+        return appendNewLine(sb.toString());
+    }
```
이 로직도 `.toString()` 용도로는 조금 과한 듯 싶기도 합니다. 
IDE의 자동생성 정도로 간소화해주세요.

polynomeer replied:
`Board::toString()` 과 `Rank::toString()`의 사용을 모두 StringUtils의 정적 메서드로 변경했습니다.

wheejuni commented on src/main/java/piece/Piece.java:
```diff
@@ -0,0 +1,104 @@
+        return this.type == type;
+    }
+
+    public void addPiecesByColor(Color color, List<Piece> piecesByColor) {
```
`color` 하나만 파라메터로 받고, 이 메소드에선 `boolean` 값을 내려준 다음에, 
리스트에 넣는 책임은 리스트를 가진 쪽에서 직접 수행해도 될 텐데요. 너무 과합니다.

polynomeer replied:
네! Rank에서 `findPiecesByColor()`가 사용중이었는데, 거기에서 이 기능을 담당하도록 했습니다. 여기에서 사용하는 `Piece::matchColor()`은 public으로 접근제한을 열게 되었습니다.

wheejuni commented on src/main/java/piece/Piece.java:
```diff
@@ -0,0 +1,104 @@
+        return matchColor(color) && matchType(type);
+    }
+
+    private boolean matchColor(Color color) {
+        return this.color == color;
+    }
+
+    private boolean matchType(Type type) {
+        return this.type == type;
+    }
```
이 메소드 두 개를 `public` 으로 오픈하면 훨씬 더 유연한 설계가 가능할 것으로 보입니다.

polynomeer replied:
우선 matchColor()는 위의 코멘트에서 오픈하게 되었습니다. matchType()도 오픈하면 좀 더 심플한 설계가 될 것 같은데, 이 부분은 좀 더 고민해보겠습니다.

wheejuni commented on src/main/java/piece/Piece.java:
```diff
@@ -0,0 +1,104 @@
+        return point;
+    }
+
+    public Color getColor() {
+        return color;
+    }
+
+    public Type getType() {
+        return type;
+    }
```
`match` 보다 `get` 이 오픈수준이 더 높은 것은 좀 어색합니다. `match` 가 더 우선 사용돼야 합니다.

polynomeer replied:
verification이나 validation 메서드가 당연히 외부 클래스에서 사용되는 경우가 많으니까 그렇겠군요. 무슨 말인지 이해했습니다.

wheejuni commented on src/main/java/piece/Piece.java:
```diff
@@ -0,0 +1,104 @@
+        return type;
+    }
+
+    public boolean isWhite() {
+        return color.equals(Color.WHITE);
+    }
+
+    public boolean isBlack() {
+        return color.equals(Color.BLACK);
+    }
```
이 두 메소드는 아래 메소드로 합칠 수 있지 않나요?

```java
public boolean isMatchingColor(Color color);

polynomeer replied:
네, `isRightColor(Color expectedColor)`이라는 메서드로 한번 합쳐보았습니다.

wheejuni commented on src/main/java/piece/Piece.java:
```diff
@@ -0,0 +1,104 @@
+        }
+    }
+
+    public double getPoint(List<Piece> pieces) {
```
이 메소드는 여기에 있으면 안될 것 같습니다. 
`Piece` 가 `List<Piece>` 를 받아서 점수를 계산하는 로직은 어색한데요. 
이 클래스엔 자기 자신의 점수를 리턴하는 정도의 로직을 배치하고, 실 점수 계산은 다른 클래스에서 이뤄져야 할 것 같네요.

polynomeer replied:
Board클래스에 `getPoint()`메서드를 추가하여 이 로직을 담당하도록 하였습니다. 여기에서는 단순히 필드 type의 point를 반환하는 메서드로 변경했습니다.

wheejuni reviewed (changes requested):
5단계까지 진행하시느라 수고 많으셨습니다. 👍 💯 
개선할 점이 좀 있는데요, 코멘트 하나하나 읽어보시고 반영해 주셨으면 좋겠습니다.
체스 과제 진행기한이 얼마 남지 않았는데 마무리가 잘 되었으면 좋겠습니다!

polynomeer committed:
refactor: Move Board::toString() into PrintUtils::showBoard()

Moved Board::toString into PrintUtils::showBoard to prevent side effect.
Changed access modifier of Rank class from default to public to use it in PrintUtils class.

polynomeer committed:
refactor: Add access modifiers to all methods and classes

Added access modifiers to all classes and methods of them to clarify meaning of access range.
Deleted public keywords in all test codes (unnecessary in JUnit5).

polynomeer committed:
refactor: Move Rank::toString into StringUtils::rankToString

Moved Rank::toString into StringUtils::rankToString.
Moved PrintUtils::showBoard into StringUtils::boardToString.
Modified test codes using Rank::toString and Board::toString.

polynomeer committed:
refactor: Deleted Piece::addPiecesByColor()

Deleted Piece::addPiecesByColor() method and added the logic in Rank::findPiecesByColor() method (which was using Piece::addPiecesByColor()).
Set the access modifier of Piece::matchColor() from private to public which Piece::addPiecesByColor() was depending on.

polynomeer committed:
refactor: Set the access modifier of Piece::matchType to public

 Set the access modifier of Piece::matchType to public to do better for structure.

polynomeer committed:
refactor: Deleted isWhite() and isBlack() in Piece class

Deleted isWhite() and isBlack() in Piece class,
and Added isRightColor() to substitute them.
Modified the test codes depending the methods.

polynomeer committed:
refactor: Deleted getPoint() in Piece class

Deleted Piece::getPoint() and Added getPoint() in Board class to put responsibility of calculating point on Board class.
Added checkAdjacentPawn() in Board class for the separation of responsibilities.
Added getPoint() and getPosition() in Piece class to use them in calculating point logic of Board class.

polynomeer committed:
refactor: Change for loop to stream based code in Board::calculatePoint

Changed for loop to stream based code in Board::calculatePoint to simplify code.

polynomeer committed:
refactor: Change for loop to stream based code in Board::findPiecesByColor

Changed for loop to stream based code in Board::findPiecesByColor to simplify code.

polynomeer commented:
리뷰에 대한 변경사항을 모두 반영하였습니다! 확인 부탁드립니다.
